src/util.js

```diff
@@ -152,39 +152,49 @@ export function createRegExpRestore () {
     return function() {
         // Now we've snapshotted some properties, escape the lastMatch string
         let esc = /[.?*+^$[\]\\(){}|-]/g,
-            lm = regExpCache.lastMatch.replace(esc, '\\$&'),
-            reg = new List();
+            lastMatch = regExpCache.lastMatch.replace(esc, '\\$&'),
+            exprStr = '';
 
         // If any of the captured strings were non-empty, iterate over them all
         if (has) {
             for (let i = 1; i <= 9; i++) {
                 let m = regExpCache['$'+i];
 
                 // If it's empty, add an empty capturing group
-                if (!m)
-                    lm = '()' + lm;
-
+                if (!m) {
+                  exprStr += '(';
+                  lastMatch = ')' + lastMatch;
+                }
                 // Else find the string in lm and escape & wrap it to capture it
                 else {
                     m = m.replace(esc, '\\$&');
-                    lm = lm.replace(m, '(' + m + ')');
+                    exprStr += lastMatch.substring(0, lastMatch.indexOf(m)) + '(';
+                    lastMatch = m + ')' + lastMatch.substring(lastMatch.indexOf(m) + m.length);
                 }
-
-                // Push it to the reg and chop lm to make sure further groups come after
-                arrPush.call(reg, lm.slice(0, lm.indexOf('(') + 1));
```

Comment:
First significant problem with the previous implementation.  `lm.indexOf('(')` is looking for the paren we added on line 170, but it could also potentially catch escaped parens that occur earlier in the string.